storage/remote/queue_manager.go

```diff
@@ -231,40 +231,30 @@ func NewQueueManager(logger log.Logger, walDir string, samplesIn *ewmaRate, cfg
 // Append queues a sample to be sent to the remote storage. Blocks until all samples are
 // enqueued on their shards or a shutdown signal is received.
 func (t *QueueManager) Append(s []tsdb.RefSample) bool {
-	type enqueuable struct {
-		ts  prompb.TimeSeries
-		ref uint64
-	}
-
-	tempSamples := make([]enqueuable, 0, len(s))
+	var ignored []int
 	t.seriesMtx.Lock()
-	for _, sample := range s {
+	for i, sample := range s {
 		// If we have no labels for the series, due to relabelling or otherwise, don't send the sample.
 		if _, ok := t.seriesLabels[sample.Ref]; !ok {
 			t.droppedSamplesTotal.Inc()
 			t.samplesDropped.incr(1)
+			ignored = append(ignored, i)
 			if _, ok := t.droppedSeries[sample.Ref]; !ok {
 				level.Info(t.logger).Log("msg", "dropped sample for series that was not explicitly dropped via relabelling", "ref", sample.Ref)
 			}
-			continue
 		}
-		tempSamples = append(tempSamples, enqueuable{
-			ts: prompb.TimeSeries{
-				Labels: t.seriesLabels[sample.Ref],
-				Samples: []prompb.Sample{
-					prompb.Sample{
-						Value:     float64(sample.V),
-						Timestamp: sample.T,
-					},
-				},
-			},
-			ref: sample.Ref,
-		})
 	}
 	t.seriesMtx.Unlock()
 
+	j := 0
 outer:
-	for _, sample := range tempSamples {
+	for i, sample := range s {
+		// Filter out any samples marked as ignored.
+		for ; j < len(ignored) && ignored[j] <= i; j++ {
```

Comment:
This smells n^2, how about using a map?

Reply:
This was still O(n) due to not resetting j during each loop. Definitely not the most obvious code I have ever written though. Also, my newest change makes it irrelevant.